### PR TITLE
feature/cheapest order on orders query

### PR DESCRIPTION
**Pull request by Jipperism**

### CODE REVIEW

File: src/utils/getTokenPriceInUSD.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+export const getTokenPriceInUSD = async (
+  chainId: ChainId,
+  tokenAddress: string,
+) => {
+  const provider = new ethers.JsonRpcProvider(getRpcUrl(chainId));
+  // This constant describes the ABI interface of the contract, which will provide the price of ETH
+  // It looks like a lot, and it is, but this information is generated when we compile the contract
+  // We need to let ethers know how to interact with this contract.
+  const aggregatorV3InterfaceABI = [
```
Comment (bitbeckers) [inline src/utils/getTokenPriceInUSD.ts:17]: don't need the full abi, you can provide the fields for the call you're actually executing. Generally ABIs are stored as JSON files for availability in the whole app as a resource

File: src/utils/getTokenPriceInUSD.ts [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+      outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
+      stateMutability: "view",
+      type: "function",
+    },
+  ];
+  // The address of the contract which will provide the price of ETH
+  const feedAddress = tokenAddressToFeedAddress(chainId, tokenAddress);
+
+  if (!feedAddress) {
+    throw new Error("Token not supported");
```
Comment (bitbeckers) [inline src/utils/getTokenPriceInUSD.ts:70]: log with token address

File: src/utils/getTokenPriceInUSD.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
+import { ethers } from "ethers";
+import { getRpcUrl } from "./getRpcUrl.js";
+import {
+  ChainId,
+  currenciesByNetwork,
+  Currency,
+} from "@hypercerts-org/marketplace-sdk";
+
+export const getTokenPriceInUSD = async (
```
Comment (bitbeckers) [inline src/utils/getTokenPriceInUSD.ts:9]: this also reflects on the `cheapestOrder` fields resolution, but this needs tests. The business logic seems to check out, but unit tests with happy and unhappy flows are valuable

File: src/utils/getTokenPriceInUSD.ts [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+};
+
+type CurrencyFeeds = Record<
+  keyof (typeof currenciesByNetwork)[ChainId],
+  string
+>;
+
+const feedsPerChain: Record<ChainId, Partial<CurrencyFeeds>> = {
+  [ChainId.SEPOLIA]: {
+    ETH: "0x694AA1769357215DE4FAC081bf1f309aDC325306",
```
Comment (bitbeckers) [inline src/utils/getTokenPriceInUSD.ts:140]: when I run the API I get a 429 on the call to this address. 

"message: 'Your app has exceeded its compute units per second capacity. If you have retries enabled, you can safely ignore this message. If not, check out https://docs.alchemy.com/reference/throughput'
"

**Review by bitbeckers — COMMENTED** [review]
Good setup, thank you! See the comments for requested fixes

File: src/graphql/schemas/resolvers/hypercertResolver.ts [hunk 5]
```diff
@@ -310,16 +306,30 @@ class HypercertResolver {
...
-          const availableOrdersForFraction = validOrders.filter(
-            (order) =>
-              parseClaimOrFractionId(fraction.fraction_id).id.toString() ===
-              order.itemIds[0],
+      const totalUnitsForSale = Object.keys(ordersByFraction)
+        .map((tokenId) => {
+          const fractionId = `${chainId}-${contractAddress}-${tokenId}`;
+          const fraction = fractionsRes.data?.find(
+            (fraction) => fraction.fraction_id === fractionId,
           );
```
[the remark this answers is not in the document]
Comment (Jipperism) [reply src/graphql/schemas/resolvers/hypercertResolver.ts:323]: `fractionsRes` is a lot of fractions, not orders. By definition there should only ever be one fraction that matches. If multiple match, it still shouldn't matter as long as they have equal units, but that would indicate a bug in the indexer anyways. If multiple orders for the same fraction are found, we take the orders that has the most units for sale, as that correctly represents the total units for sale for a hypercert.

File: src/graphql/schemas/typeDefs/hypercertTypeDefs.ts [hunk 6]
```diff
@@ -43,7 +54,7 @@ class Hypercert extends HypercertBaseType {
     nullable: true,
     description: "The metadata for the hypercert as referenced by the uri",
   })
-  metadata?: Metadata;
+  declare metadata?: Metadata;
```
[the remark this answers is not in the document]
Comment (Jipperism) [reply src/graphql/schemas/typeDefs/hypercertTypeDefs.ts:57]: Because typescript/intellij told me so, as TS was throwing a warning that we were accidentally overwriting a property. This makes that intentional. The type needs to be there because otherwise graphQL cant deduce the types.

File: src/utils/getTokenPriceInUSD.ts [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    USDC: "0x7e860098F58bBFC8648a4311b374B1D669a2bc6B",
+  },
+  [ChainId.OPTIMISM]: {
+    ETH: "0xb7B9A39CC63f856b90B364911CC324dC46aC1770",
+    WETH: "0xb7B9A39CC63f856b90B364911CC324dC46aC1770",
+    DAI: "0x8dBa75e83DA73cc766A7e5a0ee71F656BAb470d6",
+    USDC: "0x16a9FA2FDa030272Ce99B29CF780dFA30361E0f3",
+  },
+  [ChainId.CELO]: {
+    // TODO: DAI on CELO not supported
```
[the remark this answers is not in the document]
Comment (Jipperism) [reply src/utils/getTokenPriceInUSD.ts:164]: Moved it to a json file. I don't see the benefits of leaving out parts of the ABI, in case we need them at some other point in time.

File: src/utils/getTokenPriceInUSD.ts (hunk 1 above)
Comment (Jipperism) [reply src/utils/getTokenPriceInUSD.ts:17]: Moved it to a json file. I don't see the benefits of leaving out parts of the ABI, in case we need them at some other point in time.

File: src/graphql/schemas/typeDefs/hypercertTypeDefs.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (bitbeckers) [reply src/graphql/schemas/typeDefs/hypercertTypeDefs.ts:57]: That's because you are overwriting a property, please review the base types

File: src/graphql/schemas/resolvers/hypercertResolver.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (bitbeckers) [reply src/graphql/schemas/resolvers/hypercertResolver.ts:323]: Thanks for explaining, not sure if I understand this:

> f multiple orders for the same fraction are found, we take the orders that has the most units for sale, as that correctly represents the total units for sale for a hypercert.

File: src/utils/getTokenPriceInUSD.ts [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+export const getTokenPricesForChain = async (chainId: ChainId) => {
+  const currencies = currenciesByNetwork[chainId];
+  const prices = await Promise.all(
+    Object.values(currencies).map(async (currency: Currency) => {
+      try {
+        const price = await getTokenPriceInUSD(chainId, currency.address);
+        return { ...currency, price };
+      } catch (error) {
+        console.error(error);
+        return { ...currency, price: 0 };
```
[the remark this answers is not in the document]
Comment (Jipperism) [reply src/utils/getTokenPriceInUSD.ts]: What do you think is the best solution here? Should we throw an error and simply not return the cheapest hypercert if we cannot fetch the current price for a specific currency? Or not consider orders for being the cheapest if we cannot determine the current token price for the currency of that order?

This would currently only be used for determining which is the cheapest, but I imagine we will need these utilities in the frontend as well, so it might be useful to think about this for a second.

File: src/utils/getTokenPriceInUSD.ts (hunk 3 above)
Comment (Jipperism) [reply src/utils/getTokenPriceInUSD.ts:10]: I added tests for `getCheapestOrder` and `getMaxUnitsForSaleInOrders`. I'm not entirely sure what to test in `getTokenPricesInUSD`, except for the line `return Number(formatUnits(roundData * BigInt(100), decimals)) / 100;`, which seems like testing the implementation for formatUnits?

File: src/utils/getTokenPriceInUSD.ts (hunk 4 above)
Comment (Jipperism) [reply src/utils/getTokenPriceInUSD.ts:99]: How did you trigger this exactly? Simply by starting up the API? Because I'm not able to reproduce it, but would like to to see if a solution is necessary.

File: src/utils/getTokenPriceInUSD.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Jipperism) [reply src/utils/getTokenPriceInUSD.ts]: I'm just throwing an error for now, which should be caught in the top level. If another solution is preferred, please let me know.

File: src/graphql/schemas/typeDefs/hypercertTypeDefs.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jipperism) [reply src/graphql/schemas/typeDefs/hypercertTypeDefs.ts:57]: Created https://github.com/hypercerts-org/hypercerts-api/issues/128.

File: src/utils/getTokenPriceInUSD.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (bitbeckers) [reply src/utils/getTokenPriceInUSD.ts]: > This would currently only be used for determining which is the cheapest, but I imagine we will need these utilities in the frontend as well, so it might be useful to think about this for a second.

This data field `cheapestOrder` might not even belong in the Graph, but let's leave it there for now so we can ship

File: src/utils/getTokenPriceInUSD.ts (hunk 4 above)
Comment (bitbeckers) [reply src/utils/getTokenPriceInUSD.ts:99]: Starting the app, but it might've been simply my API key. After merging, please monitor for errors in the Railway logs while testing out the deployed flow

**Comment by bitbeckers** [comment]
Please don't just resolve, it's hard to keep track of changes that way

File: src/utils/getTokenPriceInUSD.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Jipperism) [reply src/utils/getTokenPriceInUSD.ts]: Oops, something got lost here. My idea was to default to the USDC price feed as they are both stable coins, so it shouldn't make much of a difference. Defaulting to DAI on mainnet adds a very specific edge case for just one token, because you need to initialize the publicClient to a different chain etc. It's more messy, for little benefit but it's an option too.

File: src/utils/getTokenPriceInUSD.ts [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+  if (!currency) {
+    throw new Error("Currency not supported");
+  }
+
+  const symbol = currency.symbol as keyof (typeof currenciesByNetwork)[ChainId];
+  const feedsForChain = feedsPerChain[chainId];
+
+  if (!feedsForChain) {
+    throw new Error("Chain not supported");
```
[the remark this answers is not in the document]
Comment (Jipperism) [reply src/utils/getTokenPriceInUSD.ts:66]: Fixed

File: src/utils/getTokenPriceInUSD.ts [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+};
+
+const tokenAddressToFeedAddress = (chainId: ChainId, tokenAddress: string) => {
+  const currencies = currenciesByNetwork[chainId];
+  const currency = Object.values(currencies).find(
+    (currency) => getAddress(currency.address) === getAddress(tokenAddress),
+  );
+
+  if (!currency) {
+    throw new Error("Currency not supported");
```
[the remark this answers is not in the document]
Comment (Jipperism) [reply src/utils/getTokenPriceInUSD.ts:59]: Fixed

File: test/api/v1/orders.test.ts [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+      {
+        additionalParameters: encodeAdditionalParams({
+          minUnitsToKeep: BigInt(1),
+        }),
+      },
+    ];
+
+    const result = getMaxUnitsForSaleInOrders(orders, BigInt(5));
+
+    expect(result).to.eq(BigInt(4));
```
[the remark this answers is not in the document]
Comment (Jipperism) [reply test/api/v1/orders.test.ts:137]: Yes, I guess that's an edge case that makes sense. We normally calculate the total amount for sale by doing `unitsInFraction - unitsToKeep`, as you would be able to do multiple orders in a row to buy all available tokens, but if `maxUnits === 0` you won't be able to buy anything at all. I actually wonder if we should support orders like that in the first place, or if they should be caught in API validation as that's an invalid order IMO.

File: test/api/v1/orders.test.ts [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+  maxUnitAmount?: bigint;
+}) => {
+  return encodeAbiParameters(
+    parseAbiParameters(
+      "uint256 minUnitAmount, uint256 maxUnitAmount, uint256 minUnitsToKeep, bool sellLeftOverFraction",
+    ),
+    [minUnitAmount, maxUnitAmount, minUnitsToKeep, sellLeftOverFraction],
+  );
+};
+describe("Get cheapest order", async () => {
```
[the remark this answers is not in the document]
Comment (Jipperism) [reply test/api/v1/orders.test.ts:29]: I've moved the tests to the appropriate folders but it doesn't seem like the tests that were already there follow the same folder structure, in the way that they do in the hypercerts-indexer. Not sure if we should clean that up.

[the remark this answers is not in the document]
Comment (Jipperism) [reply test/api/v1/orders.test.ts:29]: >  hen running the tests they also apply to the git submodules, ignore the lib folder in vitest.config.ts

fixed

File: test/api/v1/orders.test.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (Jipperism) [reply test/api/v1/orders.test.ts]: Added this edge case and a test for it, the question whether we should accept these orders in the first place remains.

File: src/graphql/schemas/resolvers/hypercertResolver.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jipperism) [reply src/graphql/schemas/resolvers/hypercertResolver.ts:324]: Solved.

**Review by bitbeckers — APPROVED** [review]
